rfcs/20180824-tf-print-v2.md

```diff
@@ -0,0 +1,549 @@
+also provides long-requested functionality for both eager and session-based
+execution, such as a more meaningful Tensor summarization, support for printing
+nested data structures that contain tensors, and controllable logging levels.
+
+## Overview
+
+We introduce two methods to the public Python API: `tf.strings.format` and
+`tf.print`. (We may have to use `tf.strings.Format` or `tf.strings.fmt` instead
+of `tf.strings.format` if the built-in python `format` method causes issues with
+the naming.) These are internally backed by two C++ operators: `StringFormat`
+and `PrintV2`. The method headers & docstrings are as follows:
+
+### Python Public API Methods
+
+````python
+@tf_export("print")
+def print_v2(*inputs, **kwargs):
+  """Print the specified inputs.
+  Returns an operator that prints the specified inputs to a desired
+  output stream or logging level. The inputs may be dense or sparse Tensors,
+  primitive python objects, data structures that contain Tensors, and printable
+  python objects. Printed tensors will recursively show the first and last
+  `summarize` elements of each dimension.
+
+  With eager execution enabled and/or inside a `tf.contrib.eager.defun` this
+  operator will automatically execute, and users only need to call `tf.print`
+  without using the return value. When constructing graphs outside of a
+  `tf.contrib.eager.defun`, one must either include the returned op
+  in the input to `session.run`, or use the operator as a control dependency for
+  executed ops by specifying `with tf.control_dependencies([print_op])`.
+
+  @compatibility(python2)
+  In python 2.7, make sure to import the following:
+  `from __future__ import print_function`
+  @end_compatibility
+
+  Example:
+    Single-input usage:
+    ```python
+    tf.enable_eager_execution()
+    tensor = tf.range(10)
+    tf.print(tensor, output_stream=sys.stderr)
+    ```
+    (This prints "[0 1 2 ... 7 8 9]" to sys.stderr)
+
+    Multi-input usage:
+    ```python
+    tf.enable_eager_execution()
+    tensor = tf.range(10)
+    tf.print("tensors:", tensor, {2: tensor * 2}, output_stream=sys.stdout)
+    ```
+    (This prints "tensors: [0 1 2 ... 7 8 9] {2: [0 2 4 ... 14 16 18]}" to sys.stdout)
+
+    Usage when constructing graphs:
+    ```python
+    sess = tf.Session()
+    with sess.as_default():
+        tensor = tf.range(10)
+        print_op = tf.print("tensors:", tensor, {2: tensor * 2},
+                            output_stream=sys.stdout)
+        with tf.control_dependencies([print_op]):
+          doubled_tensor = tensor * 2
+        sess.run(doubled_tensor)
+
+    ```
+    (This prints "tensors: [0 1 2 ... 7 8 9] {2: [0 2 4 ... 14 16 18]}" to
+    sys.stdout)
+
+  Note: This op is only partially compatible with Jupyter notebooks and colabs.
+    Because it prints to the C++ standard out / standard error, this will go
+    in the notebook kernel's console output, not in the notebook cell output.
```

Comment:
Yeah, I think so too. There's unfortunately no clean solution because python notebooks don't support this very well. As I mention lower down in the doc:

    *  Python notebooks require complex capturing logic to send C++ output to notebook cells, and
        the solutions are often not totally portable across operating systems /
        C++ runtimes.
    *  Mixing C++ output and python print output requires regular, careful flushing to prevent things happening in strange orders. This definitely has portability issues.
    *   If we want to do something about this, a possible approach could be to
        have the PrintV2 kernel execute a python print if we detect a
        jupyter/colab notebook environment and that the current device has a
        python runtime, rather than using the C++ logging/printing methods.
    *   Alternatively, we could provide utilities to capture the C++
        stdout/stderr in Jupyter notebook outputs as a part of TensorFlow.
    *   With either approach, we would have to be very careful w/ device placement in distributed
        multi-machine/multi-device settings to ensure that the print device is
        the notebook kernel CPU.

There is some internal tooling for colabs that captures stderr and prints it to the cell output which we could look into releasing. But, it's not perfect and it only flushes/prints the log output once the cell finishes executing. This can be painful w/ long-running execution so we'd need to add a regular poll for flushing.